File: src/test/java/myratpackexamples/validations/ValidationsTest.java
package myratpackexamples.validations;

import io.vavr.collection.List;
import io.vavr.collection.Seq;
import io.vavr.control.Validation;
import lombok.Value;
import org.junit.jupiter.api.Test;

import java.util.NoSuchElementException;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class ValidationsTest {

    @Value
    public static class Person {
        String name;
        Integer age;
    }

    @Test
    void if_there_is_no_validation_failure_in_the_chain_there_is_no_error() {
        Validation<Seq<String>, Person> person = Validation.combine(
                Validation.<String, String>valid("Horst"), Validation.valid(30)
        ).ap(Person::new);

        //noinspection Convert2MethodRef
        Validation<Seq<String>, Person> secondStep = person.flatMap(value -> Validation.valid(value));

        assertThrows(NoSuchElementException.class, secondStep::getError);
        assertEquals(new Person("Horst", 30), secondStep.get());
    }

    @Test
    void if_there_are_validation_failures_in_the_first_stage_further_processing_is_ignored() {
        Validation<Seq<String>, Person> person = Validation.<String, String, Integer>combine(
                Validation.invalid("Name failed"), Validation.invalid("Age failed")
        ).ap(Person::new);

        Validation<Seq<String>, Person> secondStep = person.flatMap(
                ignored -> Validation.invalid(List.of("Validation in additional step failed!"))
        );

        assertEquals(List.of("Name failed", "Age failed"), secondStep.getError());
        assertThrows(NoSuchElementException.class, secondStep::get);
    }

    @Test
    void if_no_failure_in_first_stage_and_failure_in_second_stage_error_is_failure_from_second_stage() {
        Validation<Seq<String>, Person> person = Validation.combine(
                Validation.<String, String>valid("Horst"), Validation.valid(30)
        ).ap(Person::new);

        Validation<Seq<String>, Person> secondStep = person.flatMap(
                ignored -> Validation.invalid(List.of("Validation in additional step failed!"))
        );

        assertEquals(List.of("Validation in additional step failed!"), secondStep.getError());
        assertThrows(NoSuchElementException.class, secondStep::get);
    }
}

File: src/test/java/myratpackexamples/validations/ValidationsTest.kt
package myratpackexamples.validations

import io.vavr.collection.List
import io.vavr.collection.Seq
import io.vavr.control.Validation
import myratpackexamples.webpoll.ap
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import java.util.*

class ValidationsTest {

    data class Person(val name: String?, val age: Int?)

    @Test
    fun `if there is no validation failure in the chain there is no error`() {
        val person = Validation.combine(
                Validation.valid("Horst"),
                Validation.valid<String, Int>(30)
        ).ap(::Person)

        val secondStep = person.flatMap { Validation.valid<Seq<String>, Person>(it) }

        assertThrows(NoSuchElementException::class.java) { secondStep.error }
        assertEquals(Person("Horst", 30), secondStep.get())
    }

    @Test
    fun `if there are validation failures in the first stage further processing is ignored`() {
        val person = Validation.combine(
                Validation.invalid<String, String>("Name failed"),
                Validation.invalid<String, Int>("Age failed")
        ).ap(::Person)

        val secondStep = person.flatMap {
            Validation.invalid<Seq<String>, Person>(List.of("Validation in additional step failed!"))
        }

        assertEquals(List.of("Name failed", "Age failed"), secondStep.error)
        assertThrows(NoSuchElementException::class.java) { secondStep.get() }
    }

    @Test
    fun `if no failure in first stage and failure in second stage error is failure from second stage`() {
        val person = Validation.combine(
                Validation.valid("Horst"),
                Validation.valid<String, Int>(30)
        ).ap(::Person)

        val secondStep = person.flatMap {
            Validation.invalid<Seq<String>, Person>(List.of("Validation in additional step failed!"))
        }

        assertEquals(List.of("Validation in additional step failed!"), secondStep.error)
        assertThrows(NoSuchElementException::class.java) { secondStep.get() }
    }
}
